portfolio/src/main/java/com/google/sps/servlets/FollowedTagsServlet.java

```diff
@@ -42,37 +29,73 @@
 @WebServlet("/follow-tags")
 public class FollowedTagsServlet extends HttpServlet {
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    String userEmail = (String) userService.getCurrentUser().getEmail();
+    
+    Gson gson = new Gson();
 
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(LoadFollowedTags.getFollowedTags(userEmail)));
   }
-    
-  /**
-  * Converts a ServerStats instance into a JSON string using the Gson library
-  */
+
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     UserService userService = UserServiceFactory.getUserService();
-     
-    String commentType = request.getParameter(BlogConstants.TAG_PARAMETER);
-    if (commentType == null) {
+    String userEmail = (String) userService.getCurrentUser().getEmail();
+
+    String[] receivedTags = request.getParameterValues(BlogConstants.TAG_PARAMETER);
+    if (receivedTags == null || receivedTags.length == 0) {
       return;
     }
-    String email = (String) userService.getCurrentUser().getEmail();
+    List<String> receivedTagsList = new LinkedList<String>(Arrays.asList(receivedTags));
+
+    // Prevent multiple instances of a followed tag in datastore.
+    removeFollowedTags(receivedTagsList);
+
+    putTagsInDatastore(receivedTagsList, userEmail);
+
+    response.setContentType("text/html");
+    if (receivedTagsList.size() == 1) {
+      response.getWriter().println("<h2>Success! You now follow this tag.</h2>");
+    } else if (receivedTagsList.size() > 1) {
+      response.getWriter().println("<h2>Success! You now follow these tags.</h2>");
+    } else {
+      response.getWriter().println("<h2>Failed! You follow the tag(s) already.</h2>");
+    }
+  }
 
-    Entity followedTag = new Entity(BlogConstants.TAG_QUERY);
-    followedTag.setProperty("tag", commentType);
-    followedTag.setProperty("email", email);
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    datastore.put(followedTag);
+  private void putTagsInDatastore(List<String> tags, String email) {
+    for (String tag : tags) {
+      Entity followedTag = new Entity(BlogConstants.TAG_QUERY);
+      followedTag.setProperty("tag", tag);
+      followedTag.setProperty("email", email);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(followedTag);
+    }
+  }
+
+  private void removeFollowedTags(List<String> tagsReceived) {
+    UserService userService = UserServiceFactory.getUserService();
+    String userEmail = (String) userService.getCurrentUser().getEmail();
+
+    List<String> userFollowedTags = new ArrayList<String>();
+    for (FollowedTag tagObject : LoadFollowedTags.getFollowedTags(userEmail)) {
```

Comment:
Can LoadFollowedTags.getFollowedTags fail or return null? You would get some kind of unhandled exception in this for loop in that case.

Reply:
It returns an empty list if there's nothing